test/test_ts_client.c

```diff
@@ -0,0 +1,198 @@
+}
+
+void parse_uri_null(void)
+{
+    const char *uri = NULL;
+    TSUriElems params;
+    params.ts_payload = "";
+    ts_parse_uri(uri, &params);
+    TEST_ASSERT_EQUAL_INT(-1, params.ts_list_subnodes);
+    TEST_ASSERT_EQUAL_STRING(NULL, params.ts_target_node);
+    TEST_ASSERT_EQUAL_STRING(NULL, params.ts_device_id);
+    TEST_ASSERT_EQUAL_STRING("", params.ts_payload);
+}
+
+void parse_uri_with_payload(void)
+{
+    const char *uri = "someID/output";
+    TSUriElems params;
+    params.ts_payload = "{Bat_V}";
+    ts_parse_uri(uri, &params);
+    TEST_ASSERT_EQUAL_INT(1, params.ts_list_subnodes);
+    TEST_ASSERT_EQUAL_STRING("output", params.ts_target_node);
+    TEST_ASSERT_EQUAL_STRING("someID", params.ts_device_id);
+    TEST_ASSERT_EQUAL_STRING("{Bat_V}", params.ts_payload);
+}
+
+void ts_build_query_get(void)
+{
+    TSUriElems params;
+    char *query;
+    params.ts_payload = NULL;
+    params.ts_target_node = "output";
+    params.ts_list_subnodes = 1;
+    query = ts_build_query(TS_GET, &params);
+    TEST_ASSERT_EQUAL_STRING("?output\n", query);
+    free(query);
+}
+
+void ts_build_query_get_subnodes(void)
+{
+    TSUriElems params;
+    char *query;
+    params.ts_payload = NULL;
+    params.ts_target_node = "output/";
+    params.ts_list_subnodes = 0;
+    query = ts_build_query(TS_GET, &params);
+    TEST_ASSERT_EQUAL_STRING("?output/\n", query);
+    free(query);
+}
+
+void ts_build_query_get_root(void)
+{
+    TSUriElems params;
+    char *query;
+    params.ts_payload = NULL;
+    params.ts_target_node = "";
+    params.ts_list_subnodes = 0;
+    query = ts_build_query(TS_GET, &params);
+    TEST_ASSERT_EQUAL_STRING("?/\n", query);
+    free(query);
+}
+
+void ts_build_query_exec(void)
+{
+    TSUriElems params;
+    char *query;
+    params.ts_payload = NULL;
+    params.ts_target_node = "auth";
+    params.ts_list_subnodes = 1;
+    query = ts_build_query(TS_POST, &params);
+    TEST_ASSERT_EQUAL_STRING("!auth\n", query);
+    free(query);
+}
+
+void ts_build_query_post(void)
+{
+    TSUriElems params;
+    char *query;
+    params.ts_payload = "Bat_V";
+    params.ts_target_node = "pub/serial/IDs";
+    params.ts_list_subnodes = 1;
+    query = ts_build_query(TS_POST, &params);
+    TEST_ASSERT_EQUAL_STRING("+pub/serial/IDs Bat_V\n", query);
+    free(query);
+}
+
+void ts_build_query_null(void)
+{
+    TSUriElems *params = NULL;
+    char *query;
+    query = ts_build_query(TS_GET, params);
+    TEST_ASSERT_EQUAL_STRING(NULL, query);
+}
+
+void ts_build_query_false_method(void)
+{
+    TSUriElems params;
+    char *query;
+    params.ts_payload = NULL;
+    params.ts_target_node = "output";
+    params.ts_list_subnodes = 1;
+    query = ts_build_query(0, &params);
+    TEST_ASSERT_EQUAL_STRING("", query);
+    free(query);
+}
+
+void ts_build_query_payload(void)
+{
+    TSUriElems params;
+    char *query;
+    params.ts_payload = "[Bat_V, Bat_A, Bat_W, Bat_degC, BatTempExt]";
```

Comment:
This is not valid JSON. But for the purpose of this test it probably doesn't matter...

Reply:
https://libre.solar/thingset/2b_text_mode.html#read-data
Example 4 has a payload like that? 
Or do you mean the extra apostrophes?

Reply:
Yes, JSON requires to wrap strings in `"`s.